src/Juvix/Compiler/Asm/Pretty/Base.hs

```diff
@@ -181,45 +196,181 @@ instance PrettyCode Type where
+
+instance PrettyCode Field where
+  ppCode :: Member (Reader Options) r => Field -> Sem r (Doc Ann)
+  ppCode Field {..} = do
+    dr <- ppCode _fieldRef
+    ctr <- ppConstrName _fieldTag
+    return $ dr <> dot <> ctr <> lbracket <> integer _fieldOffset <> rbracket
+
+instance PrettyCode MemValue where
+  ppCode :: Member (Reader Options) r => MemValue -> Sem r (Doc Ann)
+  ppCode = \case
+    DRef dr -> ppCode dr
+    ConstrRef fld -> ppCode fld
+
+instance PrettyCode Value where
+  ppCode :: Member (Reader Options) r => Value -> Sem r (Doc Ann)
+  ppCode = \case
+    ConstInt v ->
+      return $ annotate AnnLiteralInteger (pretty v)
+    ConstBool True ->
+      return $ annotate (AnnKind KNameConstructor) (pretty ("true" :: String))
+    ConstBool False ->
+      return $ annotate (AnnKind KNameConstructor) (pretty ("false" :: String))
+    ConstString txt ->
+      return $ annotate AnnLiteralString (pretty (show txt :: String))
+    ConstUnit {} ->
+      return $ annotate (AnnKind KNameConstructor) (pretty ("unit" :: String))
+    ConstVoid {} ->
+      return $ annotate (AnnKind KNameConstructor) (pretty ("void" :: String))
+    Ref mval ->
+      ppCode mval
+
+ppCall :: Member (Reader Options) r => Doc Ann -> InstrCall -> Sem r (Doc Ann)
+ppCall call InstrCall {..} = case _callType of
+  CallFun sym -> do
+    fn <- ppFunName sym
+    return $ call <+> fn
+  CallClosure ->
+    return $ call <+> variable "$" <+> integer _callArgsNum
+
 instance PrettyCode Instruction where
-  -- TODO: properly handle the arguments
+  ppCode :: Member (Reader Options) r => Instruction -> Sem r (Doc Ann)
   ppCode = \case
-    Binop IntAdd -> return $ pretty ("add" :: String)
-    Binop IntSub -> return $ pretty ("sub" :: String)
-    Binop IntMul -> return $ pretty ("mul" :: String)
-    Binop IntDiv -> return $ pretty ("div" :: String)
-    Binop IntMod -> return $ pretty ("mod" :: String)
-    Binop IntLt -> return $ pretty ("lt" :: String)
-    Binop IntLe -> return $ pretty ("le" :: String)
-    Binop ValEq -> return $ pretty ("eq" :: String)
-    Push {} -> return $ pretty ("push" :: String)
-    Pop -> return $ pretty ("pop" :: String)
-    PushTemp -> return $ pretty ("pusht" :: String)
-    PopTemp -> return $ pretty ("popt" :: String)
-    Trace -> return $ pretty ("trace" :: String)
-    Dump -> return $ pretty ("dump" :: String)
-    Failure -> return $ pretty ("fail" :: String)
-    Prealloc {} -> return $ pretty ("prealloc" :: String)
-    AllocConstr {} -> return $ pretty ("alloc" :: String)
-    AllocClosure {} -> return $ pretty ("calloc" :: String)
-    ExtendClosure {} -> return $ pretty ("cextend" :: String)
-    Call {} -> return $ pretty ("call" :: String)
-    TailCall {} -> return $ pretty ("tcall" :: String)
-    CallClosures {} -> return $ pretty ("ccall" :: String)
-    TailCallClosures {} -> return $ pretty ("tccall" :: String)
-    Return -> return $ pretty ("ret" :: String)
+    Binop IntAdd -> return $ primitive "add"
```

Comment:
I was expecting something different here. Instead of hard-coding the keyword strings, define those elsewhere, e.g. Asm.Data.Keyword.kwIntAdd.

Reply:
Well, I can do this if you think it's better style.

Reply:
But these are not keyword, strictly speaking, from the point of view of the parser.

Reply:
I moved the hardcoded string literals to Juvix.Extra.Strings